### DISCUSSION TITLE
IKVM.Runtime.InternalException : Could not load libjvm

### DISCUSSION THREAD

**Discussion by san-coder1** [sergey-tihon/Stanford.NLP.NET#134] [Q&A]
I am running my application from Azure Functions. It builds the .dll needed and places it in the bin folder. `Assembly.LoadFile(modelsAssemblyPath);` successfully loads the dll. 

It fails with following error during `var pipeline = new StanfordCoreNLP(props);`

```
System.TypeInitializationException : The type initializer for '<Module>' threw an exception. ---> System.TypeInitializationException : The type initializer for '<Module>' threw an exception. ---> System.TypeInitializationException : The type initializer for 'IKVM.Runtime.LibJava' threw an exception. ---> System.TypeInitializationException : The type initializer for 'IKVM.Runtime.LibJvm' threw an exception. ---> IKVM.Runtime.InternalException : Could not load libjvm.
   at IKVM.Runtime.LibJvm..ctor()
   at IKVM.Runtime.LibJvm..cctor()
   End of inner exception
   at IKVM.Runtime.LibJava..ctor()
   at IKVM.Runtime.LibJava..cctor()
   End of inner exception
   at IKVM.Runtime.BootstrapClassLoader..ctor(RuntimeContext context)
   at IKVM.Runtime.RuntimeClassLoaderFactory.GetBootstrapClassLoader()
   at IKVM.Runtime.RuntimeClassLoaderFactory.GetClassLoaderWrapper(ClassLoader javaClassLoader)
   at IKVM.Runtime.JNI.JNIFrame.GetFuncPtr(Object callerID,String clazz,String name,String sig)
   at java.lang.System.initProperties(Properties props)
   at java.lang.System.initializeSystemClass()
   at __<MethodAccessor>__System__initializeSystemClass()
   at IKVM.Runtime.Accessors.Java.Lang.SystemAccessor.InvokeInitializeSystemClass()
   at IKVM.Runtime.JVM.Init()
   at IKVM.Runtime.RuntimeInit.Init()
   at .cctor()
   End of inner exception
   at ikvm.runtime.Startup.init()
   at .cctor()
   End of inner exception
```
It works perfectly fine from my local machine, just not on Azure Function.

**Comment by sergey-tihon** [comment]
I recommend opening an issue in the IKVM project - https://github.com/ikvmnet/ikvm/issues

or take a look at the source code
https://github.com/ikvmnet/ikvm/blob/81811efe971c7a2d2aee69ed336d38de95e8e5fa/src/IKVM.Runtime/LibJVM.cs#L136-L137

and take a look at probing location
https://github.com/ikvmnet/ikvm/blob/81811efe971c7a2d2aee69ed336d38de95e8e5fa/src/IKVM.Runtime/JVM.Properties.cs#L119

- try to set `ikvm.home` or `ikvm.home.root`
- check the values of `AppContext.BaseDirectory` and `AppDomain.CurrentDomain.BaseDirectory` inside Azure Functions runtime

**Reply by san-coder1** [reply]
What should I set in the ikvm.home or ikvm.home.root.
I see ikvm folder in and jvm.dll in two places in my Azure Function directory.

1. c:\home\site\wwwroot\ikvm
2. c:\home\site\wwwroot\bin\ikvm

c:\home\site\wwwroot\ikvm\bin only contains ikvm.properties file
![image](https://github.com/user-attachments/assets/84ceaf5b-2e8d-4f92-9f61-33fb74d196e9)

c:\home\site\wwwroot\bin\ikvm\win-x64\bin> contains the jvm.dll
![image](https://github.com/user-attachments/assets/417b6fbd-b4fe-46e5-903a-e90db23c86ad)

**Reply by sergey-tihon** [reply]
look like it should point to the folder with props file and there should be bin sub-foder with the dlls inside
```sharp
 if ((Handle = NativeLibrary.Load(Path.Combine(JVM.Properties.HomePath, "bin", NativeLibrary.MapLibraryName("jvm")))) == null)
                throw new InternalException("Could not load libjvm.");
```

**Reply by san-coder1** [reply]
Here are the locations of ikvm.properties and jvm.dll. Which is the properties file that I need to modify and with which location?

Location of file ikvm.properties:

\home\site\wwwroot\ikvm.properties
\home\site\wwwroot\ikvm\win-x64\bin\ikvm.properties

Location of file jvm.dll
\home\site\wwwroot\ikvm\win-x64\bin\jvm.dll
\home\site\wwwroot\bin\ikvm\win-x64\bin\jvm.dll

**Reply by sergey-tihon** [reply]
Sorry, I do not know the answer better than to look at how IKVM loads dlls or contact library authors.
